README.md
使用　基于canvas撸的一个 Flutter头像裁剪控件

* TODO:单指移动，双指移动、缩放、旋转
* TODO:左键移动
* TODO:滚轮旋转
* TODO:左键＋滚轮缩放

![输入图片说明](./index.webp?raw=true "在这里输入图片标题")
输入图片说明

属性 backBoxSize 背景方格大小 默认值：10

backBoxColor0 背景方格颜色0 默认值：Colors.grey

backBoxColor1　　背景方格颜色1 默认值：Colors.white

outWidth 输出图片宽度 默认值：256

outHeight 输出图片高度 默认值：256

maskColor 蒙板颜色 默认值：#00000080

maskPadding 蒙板内边距 默认值：20

lineWidth 预览框线宽 默认值：3

lineColor 预览框颜色 默认值：Colors.white

isArc 是否是圆形 默认值：true

round 预览框圆色 默认值：8

limitations 限制位置和尺寸 默认值：false

image 输入图片源 


方法

outImage() 返回裁剪后的图片


使用
在pubspec.yaml文件中添加
```yaml
dependencies:
  head_image_cropper: ^3.0.0
```


样例：
```dart
import 'dart:io';
import 'dart:typed_data';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:head_image_cropper/head_image_cropper.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  var _controller = CropperController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        actions: <Widget>[
          FlatButton(
            child: Text("保存"),
            onPressed: () {
              _controller.outImage()?.then((image) async {
                //保存或上传代码
                var bytes =
                    (await (image.toByteData(format: ImageByteFormat.png)))
                        .buffer
                        .asUint8List();

                Navigator.of(context)
                    .push(MaterialPageRoute(builder: (context) {
                  return ShowImage(
                    data: bytes,
                  );
                }));
//                  File("path").writeAsBytesSync(bytes);
              });
            },
          )
        ],
      ),
      body: Container(
        padding: EdgeInsets.all(50),
        child: CropperImage(
          AssetImage("images/test.webp"),
          controller: _controller,
        ),
      ),
    );
  }
}

class ShowImage extends StatefulWidget {
  final Uint8List data;

  const ShowImage({Key key, this.data}) : super(key: key);

  @override
  _ShowImageState createState() => _ShowImageState();
}

class _ShowImageState extends State<ShowImage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("裁剪效果"),
      ),
      body: Image.memory(widget.data),
    );
  }
}

```


CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.

## [1.0.0] - TODO: Add release date.

* TODO: 简化代码。修改BUG.

## [2.0.0] - TODO: Add release date.

* TODO: 重写代码.
* TODO:添加鼠标支持
* TODO:左键移动
* TODO:滚轮旋转
* TODO:左键＋滚轮缩放
* TODO:添加尺寸限制功能

## [2.0.3] - TODO: Add release date.

* 修改BUG.## 

[2.0.4] - TODO: Add release date.

* 格式化代码.

[2.0.5] - TODO: Add release date.

* 修改WEB端的BUG.

[2.0.6] - 2020/05/11

* 修改Web端不能输出bytes的BUG


[2.0.8] - 2020/06/14

* 修改IOS不能使用的问题
* 完成所有平台测试

[3.0.0] - 2020/11/29

* 更新Flutter SDK to V1.25.0
* 添加CropperController输出图像
* 删除Element输出图像

[3.0.1] - 2021/03/30

* 修改BUG


[3.1.0] - 2021/07/13

* 添加空安全支持
* Add null-safet support



example/lib/main.dart
import 'dart:typed_data';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:head_image_cropper/head_image_cropper.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Home(),
    );
  }
}

class Home extends StatefulWidget {
  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  var _controller = CropperController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        actions: <Widget>[
          TextButton(
            child: Text("保存"),
            style: ButtonStyle(foregroundColor: MaterialStateProperty.all(Colors.white)),
            onPressed: () {
              _controller.outImage().then((image) async {
                //保存或上传代码
                var bytes =
                    (await (image.toByteData(format: ImageByteFormat.png)))!
                        .buffer
                        .asUint8List();

                Navigator.of(context)
                    .push(MaterialPageRoute(builder: (context) {
                  return ShowImage(
                    data: bytes,
                  );
                }));
//                  File("path").writeAsBytesSync(bytes);
              });
            },
          )
        ],
      ),
      body: Container(
        padding: EdgeInsets.all(50),
        child: CropperImage(
          AssetImage("images/test.webp"),
          controller: _controller,
        ),
      ),
    );
  }
}

class ShowImage extends StatefulWidget {
  final Uint8List data;

  const ShowImage({Key? key, required this.data}) : super(key: key);

  @override
  _ShowImageState createState() => _ShowImageState();
}

class _ShowImageState extends State<ShowImage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("裁剪效果"),
      ),
      body: Image.memory(widget.data),
    );
  }
}


lib/cropper_image.dart
library head_image_cropper;

import 'dart:ui';

import 'package:flutter/material.dart';
import 'dart:ui' as sky show Image;

class CropperImage extends StatefulWidget {
  static const Color defualtMaskColor = Color.fromARGB(160, 0, 0, 0);

  CropperImage(
    this.image, {
    Key key,
    this.limitations = true,
    this.isArc = false,
    this.backBoxSize = 10.0,
    this.backBoxColor0 = Colors.grey,
    this.backBoxColor1 = Colors.white,
    this.maskColor = defualtMaskColor,
    this.lineColor = Colors.white,
    this.lineWidth = 3,
    this.outWidth = 256.0,
    this.outHeight = 256.0,
    this.maskPadding = 20.0,
    this.round = 8.0,
  }) : super(key: key);

  ImageProvider image;
  bool limitations;
  bool isArc;
  double backBoxSize;
  Color backBoxColor0;
  Color backBoxColor1;
  Color maskColor;
  Color lineColor;
  double lineWidth;
  double outWidth;
  double outHeight;
  double maskPadding;
  double round;

  @override
  CropperImageState createState() => CropperImageState();
}

class CropperImageState extends State<CropperImage> {
  _Param _param = _Param();
  Size _size = Size.zero;
  ImageStream _imageStream;
  ImageInfo _imageInfo;
  bool _isListeningToStream = false;
  double _scale;
  double _rotation;
  double _dx;
  double _dy;

  ImageStreamListener _imageStreamListener;

  @override
  void didChangeDependencies() {
    _resolveImage();

    if (TickerMode.of(context))
      _listenToStream();
    else
      _stopListeningToStream();

    super.didChangeDependencies();
  }

  @override
  void didUpdateWidget(CropperImage oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.image != oldWidget.image) _resolveImage();
  }

  @override
  void reassemble() {
    _resolveImage(); // in case the image cache was flushed
    super.reassemble();
  }

  void _resolveImage() {
    final ImageStream newStream = widget.image.resolve(createLocalImageConfiguration(
      context,
    ));
    assert(newStream != null);
    _updateSourceStream(newStream);
  }

  void _handleImageChanged(ImageInfo imageInfo, bool synchronousCall) {
    setState(() {
      _param = _Param(
        limitations: widget.limitations,
        isArc: widget.isArc,
        backBoxColor0: widget.backBoxColor0,
        backBoxColor1: widget.backBoxColor1,
        backBoxSize: widget.backBoxSize,
        maskColor: widget.maskColor,
        lineColor: widget.lineColor,
        lineWidth: widget.lineWidth,
        outWidth: widget.outWidth,
        outHeight: widget.outHeight,
        maskPadding: widget.maskPadding,
        round: widget.round,
      );
      _imageInfo = imageInfo;
    });
  }

  // Update _imageStream to newStream, and moves the stream listener
  // registration from the old stream to the new stream (if a listener was
  // registered).
  void _updateSourceStream(ImageStream newStream) {
    if (_imageStream?.key == newStream?.key) return;

    if (_isListeningToStream) _imageStream.removeListener(_imageStreamListener);

    _imageStream = newStream;
    if (_isListeningToStream) _imageStream.addListener(_imageStreamListener);
  }

  void _listenToStream() {
    if (_isListeningToStream) return;
    _imageStream.addListener(_imageStreamListener);
    _isListeningToStream = true;
  }

  void _stopListeningToStream() {
    if (!_isListeningToStream) return;
    _imageStream.removeListener(_imageStreamListener);
    _isListeningToStream = false;
  }

  @override
  void dispose() {
    assert(_imageStream != null);
    _stopListeningToStream();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onScaleStart: (details) {
        _scale = 1;
        _rotation = 0;
        _dx = details.focalPoint.dx;
        _dy = details.focalPoint.dy;
      },
      onScaleUpdate: (details) {
        setState(() {
          _param.scale *= (details.scale / _scale);
          _scale = details.scale;

          _param.rotate += (details.rotation - _rotation);
          _rotation = details.rotation;

          _param.drawX += details.focalPoint.dx - _dx;
          _dx = details.focalPoint.dx;

          _param.drawY += details.focalPoint.dy - _dy;
          _dy = details.focalPoint.dy;
          print(details.toString());
        });
      },
      child: CustomPaint(
        size: _size,
        painter: _CropperImagePainter(_imageInfo, _param),
      ),
    );
  }

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((callback) {
      setState(() {
        _size = MediaQuery.of(context).size;
      });
    });

    _imageStreamListener = ImageStreamListener(_handleImageChanged);
  }

  Future<sky.Image> outImage() {
    var recorder = PictureRecorder();
    var canvas = Canvas(recorder, Rect.fromLTRB(0, 0, _param.outWidth, _param.outHeight));

    if (null != _imageInfo) {
      var scale = widget.outHeight / (_param.bottom - _param.top);
      canvas.translate(_param.outWidth / 2 + _param.drawX * scale, _param.outHeight / 2 + _param.drawY * scale);

      canvas.rotate(_param.rotate);
      canvas.scale(_param.scale * scale);
      canvas.drawImage(
          _imageInfo.image,
          Offset(
            -_imageInfo.image.width / 2,
            -_imageInfo.image.height / 2,
          ),
          Paint());
    }

    return recorder.endRecording().toImage(_param.outWidth.toInt(), _param.outHeight.toInt());
  }
}

class _Param {
  bool limitations = true;
  bool isArc = false;
  double backBoxSize = 10.0;
  Color backBoxColor0 = Colors.grey;
  Color backBoxColor1 = Colors.white;
  Color maskColor = Color.fromARGB(80, 0, 0, 0);
  Color lineColor = Colors.white;
  double lineWidth = 3;
  double outWidth = 256.0;
  double outHeight = 256.0;
  double maskPadding = 20.0;
  double round = 8.0;

  _Param(
      {this.limitations,
      this.isArc,
      this.backBoxSize,
      this.backBoxColor0,
      this.backBoxColor1,
      this.maskColor,
      this.lineColor,
      this.lineWidth,
      this.outWidth,
      this.outHeight,
      this.maskPadding,
      this.round});

  double scale = 0;
  double centerX;
  double centerY;
  double drawX = 0;
  double drawY = 0;
  double bottom;
  double left;
  double right;
  double top;
  double rotate = 0;
}

class _CropperImagePainter extends CustomPainter {
  ImageInfo image = null;
  _Param _param;

  _CropperImagePainter(this.image, this._param);

  @override
  void paint(Canvas canvas, Size size) {
    if (size == Size.zero) {
      return;
    }

    canvas.drawColor(Colors.blue, BlendMode.clear);
    _onPadding(size);
    _createBack(canvas, size);
    if (null != image) {
      _onPosition();
      canvas.save();
      canvas.translate(_param.centerX + _param.drawX, _param.centerY + _param.drawY);
      canvas.rotate(_param.rotate);
      canvas.scale(_param.scale);
      canvas.drawImage(
          image.image,
          Offset(
            -image.image.width / 2,
            -image.image.height / 2,
          ),
          Paint());
      canvas.restore();
    }

    _craeteMask(canvas, size);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }

  _createBack(Canvas canvas, Size size) {
    for (double y = 0; y < size.height; y += _param.backBoxSize) {
      var color = (0 == (y / _param.backBoxSize) % 2) ? _param.backBoxColor0 : _param.backBoxColor1;

      for (double x = 0; x < size.width; x += _param.backBoxSize) {
        canvas.drawRect(Rect.fromLTRB(x, y, x + _param.backBoxSize, y + _param.backBoxSize),
            Paint()..color = (color = color == _param.backBoxColor1 ? _param.backBoxColor0 : _param.backBoxColor1));
      }
    }
  }

  _craeteMask(Canvas canvas, Size size) {
    if (_param.isArc) {
      canvas.drawPath(
          Path()
            ..moveTo(0, 0)
            ..lineTo(0, size.height)
            ..lineTo(size.width, size.height)
            ..lineTo(size.width, 0)
            ..addOval(Rect.fromLTWH(_param.left, _param.top, _param.right, _param.bottom))
            ..close(),
          Paint()
            ..color = _param.maskColor
            ..style = PaintingStyle.fill);

      canvas.drawPath(
          Path()
            ..addOval(Rect.fromLTWH(_param.left, _param.top, _param.right, _param.bottom))
            ..close(),
          Paint()
            ..color = _param.lineColor
            ..strokeWidth = _param.lineWidth
            ..style = PaintingStyle.stroke);
    } else {
      canvas.drawPath(
          Path()
            ..moveTo(0, 0)
            ..lineTo(0, size.height)
            ..lineTo(size.width, size.height)
            ..lineTo(size.width, 0)
            ..addRRect(RRect.fromLTRBXY(_param.left, _param.top, _param.right, _param.bottom, _param.round, _param.round))
            ..close(),
          Paint()
            ..color = _param.maskColor
            ..style = PaintingStyle.fill);

      canvas.drawPath(
          Path()
            ..addRRect(RRect.fromLTRBXY(_param.left, _param.top, _param.right, _param.bottom, _param.round, _param.round))
            ..close(),
          Paint()
            ..color = _param.lineColor
            ..strokeWidth = _param.lineWidth
            ..style = PaintingStyle.stroke);
    }
  }

  _onPadding(Size size) {
    var fw = size.width / _param.outWidth;
    var fh = size.height / _param.outHeight;
    if (fw > fh) {
      fw = fh;
    }
    var width = _param.outWidth * fw / 2 - _param.maskPadding;
    var height = _param.outHeight * fw / 2 - _param.maskPadding;
    _param.centerX = size.width / 2;
    _param.centerY = size.height / 2;
    _param.left = _param.centerX - width;
    _param.right = _param.centerX + width;
    _param.top = _param.centerY - height;
    _param.bottom = _param.centerY + height;
  }

  _onPosition() {
    if (5 < _param.scale) {
      _param.scale = 5;
    }

    var fw = (_param.right - _param.left) / image.image.width;
    var fh = (_param.bottom - _param.top) / image.image.height;
    if (fw < fh) {
      fw = fh;
    }
    if (_param.scale < fw) {
      _param.scale = fw;
    }
    //  TODO 限制
    // drawX
    // drawY
    // rotate
  }
}


lib/src/cropper_image_web_out.dart
import 'dart:async';
import 'dart:html';
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'cropper_image_out.dart' as out;

Future<ui.Image> outImage({
  required ui.Image image,
  required double outWidth,
  required double outHeight,
  required double bottom,
  required double top,
  required double drawX,
  required double drawY,
  required double rotate1,
  required double scale,
}) {
  if (!(image is ImageElement)) {
    return out.outImage(
      image: image,
      outWidth: outWidth,
      outHeight: outHeight,
      bottom: bottom,
      top: top,
      drawX: drawX,
      drawY: drawY,
      rotate1: rotate1,
      scale: scale,
    );
  }
  var canvas =
      CanvasElement(width: outWidth.toInt(), height: outHeight.toInt());
    var temp = outHeight / (bottom - top);

    canvas.context2D
        .translate(outWidth / 2 + drawX * temp, outHeight / 2 + drawY * temp);
    canvas.context2D.rotate(rotate1);
    canvas.context2D.scale(scale * temp, scale * temp);
    canvas.context2D.drawImage(
        (image as dynamic).imgElement, -image.width / 2, -image.height / 2);

  return Future.value((HtmlImage(canvas)));
}

class HtmlImage implements ui.Image {
  CanvasElement element;

  HtmlImage(this.element);

  @override
  void dispose() {}

  @override
  int get height => element.height!;

  @override
  Future<ByteData> toByteData(
      {ui.ImageByteFormat format = ui.ImageByteFormat.png}) async {
    var completer = new Completer<ByteData>();

    final out = new FileReader();
    out.onLoadEnd.listen((event) {
      completer.complete((out.result as Uint8List).buffer.asByteData());
    });
    out.onError.listen((event) {
      completer.completeError(event);
    });
    out.readAsArrayBuffer(await element.toBlob());
    return completer.future;
  }

  int get width => element.width!;

  @override
  ui.Image clone() {
    // TODO: implement clone
    throw UnimplementedError();
  }

  @override
  bool get debugDisposed => throw UnimplementedError();

  @override
  List<StackTrace> debugGetOpenHandleStackTraces() {
    // TODO: implement debugGetOpenHandleStackTraces
    throw UnimplementedError();
  }

  @override
  bool isCloneOf(ui.Image other) {
    return false;
  }
}


lib/head_image_cropper.dart
library head_image_cropper;

import 'dart:async';
import 'dart:math' as math;
import 'dart:ui';
import 'dart:ui' as ui show Image;
import 'src/cropper_image_out.dart'
    if (dart.library.html) 'src/cropper_image_web_out.dart' as imgOut;

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

const Color _defualtMaskColor = Color.fromARGB(160, 0, 0, 0);

class CropperController {
  CropperImageElement? _element;

  Future<ui.Image> outImage() {
    return _element!._outImage();
  }
}

///图像裁剪，适用于头像裁剪和输出固定尺寸的图片裁剪
class CropperImage extends RenderObjectWidget {
  CropperImage(
    this.image, {
    Key? key,
    this.controller,
    this.limitations = true,
    this.isArc = false,
    this.backBoxSize = 10.0,
    this.backBoxColor0 = Colors.grey,
    this.backBoxColor1 = Colors.white,
    this.maskColor = _defualtMaskColor,
    this.lineColor = Colors.white,
    this.lineWidth = 3,
    this.outWidth = 256.0,
    this.outHeight = 256.0,
    this.maskPadding = 20.0,
    this.round = 8.0,
  }) : super(key: key);

  ///  image 输入图片源
  final ImageProvider image;

  ///  limitations 限制位置和尺寸 默认值：false 当为true 是不能旋转图像
  final bool limitations;

  ///  isArc 是否是圆形 默认值：false
  final bool isArc;

  /// backBoxSize 背景方格大小 默认值：10
  final double backBoxSize;

  ///  backBoxColor0 背景方格颜色0 默认值：Colors.grey
  final Color backBoxColor0;

  ///  backBoxColor1　　背景方格颜色1 默认值：Colors.white
  final Color backBoxColor1;

  ///  maskColor 蒙板颜色 默认值：#00000080
  final Color maskColor;

  ///  lineColor 预览框颜色 默认值：Colors.white
  final Color lineColor;

  ///  lineWidth 预览框线宽 默认值：3
  final double lineWidth;

  ///  outWidth 输出图片宽度 默认值：256
  final double outWidth;

  ///  outHeight 输出图片高度 默认值：256
  final double outHeight;

  ///  maskPadding 蒙板内边距 默认值：20
  final double maskPadding;

  ///round 预览框圆角 默认值：8
  final double round;

  final CropperController? controller;

  @override
  CropperImageElement createElement() {
    return CropperImageElement(this);
  }

  @override
  CropperImageRender createRenderObject(BuildContext context) {
    return CropperImageRender()
      ..limitations = limitations
      ..isArc = isArc
      ..backBoxColor0 = backBoxColor0
      ..backBoxColor1 = backBoxColor1
      ..backBoxSize = backBoxSize
      ..maskColor = maskColor
      ..lineColor = lineColor
      ..lineWidth = lineWidth
      ..outWidth = outWidth
      ..outHeight = outHeight
      ..maskPadding = maskPadding
      ..round = round;
  }

  @override
  void updateRenderObject(
      BuildContext context, CropperImageRender renderObject) {
    renderObject
      ..limitations = limitations
      ..isArc = isArc
      ..backBoxColor0 = backBoxColor0
      ..backBoxColor1 = backBoxColor1
      ..backBoxSize = backBoxSize
      ..maskColor = maskColor
      ..lineColor = lineColor
      ..lineWidth = lineWidth
      ..outWidth = outWidth
      ..outHeight = outHeight
      ..maskPadding = maskPadding
      ..round = round;
    renderObject.markNeedsPaint();
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<ImageProvider>('image', image));
    properties.add(DiagnosticsProperty<bool>('limitations', limitations));
    properties.add(DiagnosticsProperty<bool>('isArc', isArc));
    properties.add(DoubleProperty('backBoxSize', backBoxSize));
    properties.add(ColorProperty('backBoxColor0', backBoxColor0));
    properties.add(ColorProperty('backBoxColor1', backBoxColor1));
    properties.add(ColorProperty('maskColor', maskColor));
    properties.add(ColorProperty('lineColor', lineColor));
    properties.add(DoubleProperty('lineWidth', lineWidth));
    properties.add(DoubleProperty('outWidth', outWidth));
    properties.add(DoubleProperty('outHeight', outHeight));
    properties.add(DoubleProperty('maskPadding', maskPadding));
    properties.add(DoubleProperty('round', round));
  }
}

class CropperImageElement extends RenderObjectElement {
  ImageProvider? _image;

  CropperImageElement(CropperImage widget) : super(widget);

  @override
  CropperImageRender get renderObject =>
      super.renderObject as CropperImageRender;

  @override
  CropperImage get widget => super.widget as CropperImage;

  @override
  void forgetChild(Element child) {
   super.forgetChild(child);
  }

  void _resolveImage() {
    if (null == _image) {
      return;
    }
    final ImageStream stream =
        _image!.resolve(createLocalImageConfiguration(this));
    var listener;
    listener = ImageStreamListener((image, synchronousCall) {
      renderObject.image = image.image;
      stream.removeListener(listener);
    }, onError: (exception, stackTrace) {
      stream.removeListener(listener);
    });
    stream.addListener(listener);
  }

  @override
  void update(CropperImage newWidget) {
    super.update(newWidget);
    if (_image != newWidget.image) {
      _image = widget.image;
      _resolveImage();
    }
    newWidget.controller?._element = this;
  }

  @override
  void mount(Element? parent, dynamic newSlot) {
    super.mount(parent, newSlot);
    _image = widget.image;
    widget.controller?._element = this;
    _resolveImage();
  }

  Future<ui.Image> _outImage() {
    return imgOut.outImage(
      image: renderObject.image!,
      outWidth: widget.outWidth,
      outHeight: widget.outHeight,
      bottom: renderObject.bottom!,
      top: renderObject.top!,
      drawX: renderObject.drawX,
      drawY: renderObject.drawY,
      rotate1: renderObject.rotate1,
      scale: renderObject.scale,
    );
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<ImageProvider>('_image', _image));
  }
}

class Pointer {
  final int? device;
  final double? dx;
  final double? dy;

  Pointer({this.device, this.dx, this.dy});

  @override
  String toString() {
    return 'Pointer{device: $device, dx: $dx, dy: $dy}';
  }
}

class CropperImageRender extends RenderProxyBox {
  ui.Image? _image;
  bool _limitations = true;

  set limitations(bool value) {
    _limitations = value;
    if (_limitations) {
      rotate1 = 0;
    }
  }

  bool get limitations => _limitations;

  bool isArc = false;
  double backBoxSize = 10.0;
  Color backBoxColor0 = Colors.grey;
  Color backBoxColor1 = Colors.white;
  Color maskColor = Color.fromARGB(80, 0, 0, 0);
  Color lineColor = Colors.white;
  double lineWidth = 3;
  double outWidth = 256.0;
  double outHeight = 256.0;
  double maskPadding = 20.0;
  double round = 8.0;

  double scale = 0;
  double? centerX;
  double? centerY;
  double drawX = 0;
  double drawY = 0;
  double? bottom;
  double? left;
  double? right;
  double? top;
  double rotate1 = 0;

  Pointer? _old1, _old2, _new1, _new2;

  set image(ui.Image? image) {
    _image = image;
    markNeedsPaint();
  }

  ui.Image? get image => _image;

  @override
  void handleEvent(PointerEvent event, BoxHitTestEntry entry) {
    super.handleEvent(event, entry);
    if (event is PointerScrollEvent) {
      handleScrollEvent(event);
    } else if (event is PointerDownEvent) {
      handleDownEvent(event);
    } else if (event is PointerMoveEvent) {
      handleMoveEvent(event);
    } else if (event is PointerUpEvent) {
      handleUpEvent(event);
    }
  }

  void handleScrollEvent(PointerScrollEvent event) {
    if (null == _old1 && null == _old2) {
      if (event.scrollDelta.dy < 0) {
        scale -= 0.05;
      } else if (event.scrollDelta.dy > 0) {
        scale += 0.05;
      }
    } else if (!limitations) {
      if (event.scrollDelta.dy < 0) {
        rotate1 -= 0.05;
      } else if (event.scrollDelta.dy > 0) {
        rotate1 += 0.05;
      }
    }
    markNeedsPaint();
  }

  void handleDownEvent(PointerDownEvent event) {
    if (null == _old1 && _old2?.device != event.device) {
      _old1 = Pointer(
          device: event.device, dx: event.position.dx, dy: event.position.dy);
    } else if (null == _old2 && _old1!.device != event.device) {
      _old2 = Pointer(
          device: event.device, dx: event.position.dx, dy: event.position.dy);
    }
  }

  void handleMoveEvent(PointerMoveEvent event) {
    if (_old1?.device == event.device) {
      _new1 = Pointer(
          device: event.device, dx: event.position.dx, dy: event.position.dy);
    } else if (_old2?.device == event.device) {
      _new2 = Pointer(
          device: event.device, dx: event.position.dx, dy: event.position.dy);
    }

    if (null != _old1 && null != _old2 && null != _new1 && null != _new2) {
      var newLine = math.sqrt(
          math.pow(_new1!.dx! - _new2!.dx!, 2) + math.pow(_new1!.dy! - _new2!.dy!, 2));
      var oldLine = math.sqrt(
          math.pow(_old1!.dx! - _old2!.dx!, 2) + math.pow(_old1!.dy! - _old2!.dy!, 2));
      this.scale *= (newLine / oldLine);

      this.drawX += ((_new1!.dx! - _old1!.dx!) + (_new2!.dx! - _old2!.dx!)) / 2;
      this.drawY += ((_new1!.dy! - _old1!.dy!) + (_new2!.dy! - _old2!.dy!)) / 2;

      if (!limitations) {
        var k1 = (_old1!.dx! - _old2!.dx!) / (_old1!.dy! - _old2!.dy!);
        var k2 = (_new1!.dx! - _new2!.dx!) / (_new1!.dy! - _new2!.dy!);

        var temp = ((k2 - k1) / (1 + k1 * k2) * math.pi / 2);
        if (!temp.isNaN) {
          this.rotate1 -= temp;
        }
      }
      markNeedsPaint();
    } else if ((null != _old1 && null != _new1) ||
        (null != _old2 && null != _new2)) {
      this.drawX += ((_new1 ?? _new2)!.dx! - (_old1 ?? _old2)!.dx!);
      this.drawY += ((_new1 ?? _new2)!.dy! - (_old1 ?? _old2)!.dy!);
      markNeedsPaint();
    }
    if (_old1?.device == event.device) {
      _old1 = _new1;
    } else if (_old2?.device == event.device) {
      _old2 = _new2;
    }
  }

  void handleUpEvent(PointerUpEvent event) {
    if (_old1?.device == event.device) {
      _old1 = _new1 = null;
    } else if (_old2?.device == event.device) {
      _old2 = _new2 = null;
    }
  }

  @override
  bool hitTestSelf(Offset position) {
    return true;
  }

  @override
  void performResize() {
    size = constraints.biggest;
  }

  @override
  void performLayout() {}

  @override
  bool get sizedByParent => true;

  //////////////////////////////////////////////////////////////////////////////////////////////

  @override
  void paint(PaintingContext context, Offset offset) {
    if (size == Size.zero) {
      return;
    }
    var canvas = context.canvas;
    canvas.save();
    canvas.translate(offset.dx, offset.dy);
    canvas.clipRect(Rect.fromLTWH(0, 0, size.width, size.height));
//    canvas.drawColor(Colors.blue, BlendMode.color);
    _onPadding(size);
    _createBack(canvas, size);
    if (null != _image) {
      _onPosition();
      canvas.save();
      canvas.translate(centerX! + drawX, centerY! + drawY);
      canvas.rotate(rotate1);
      canvas.scale(scale);
      canvas.drawImage(
          _image!, Offset(-_image!.width/ 2, -_image!.height/ 2), Paint());
      canvas.restore();
    }

    _craeteMask(canvas, size);
    canvas.restore();
  }

  _createBack(Canvas canvas, Size size) {
    for (double y = 0; y < size.height; y += backBoxSize) {
      var color = (0 == (y / backBoxSize) % 2) ? backBoxColor0 : backBoxColor1;

      for (double x = 0; x < size.width; x += backBoxSize) {
        canvas.drawRect(
            Rect.fromLTRB(x, y, x + backBoxSize, y + backBoxSize),
            Paint()
              ..color = (color =
                  color == backBoxColor1 ? backBoxColor0 : backBoxColor1));
      }
    }
  }

  _craeteMask(Canvas canvas, Size size) {
    if (isArc) {
      canvas.drawPath(
          Path()
            ..moveTo(0, 0)
            ..lineTo(0, size.height)
            ..lineTo(size.width, size.height)
            ..lineTo(size.width, 0)
            ..addOval(Rect.fromLTRB(left!, top!, right!, bottom!))
            ..close(),
          Paint()
            ..color = maskColor
            ..style = PaintingStyle.fill);

      canvas.drawPath(
          Path()
            ..addOval(Rect.fromLTRB(left!, top!, right!, bottom!))
            ..close(),
          Paint()
            ..color = lineColor
            ..strokeWidth = lineWidth
            ..style = PaintingStyle.stroke);
    } else {
      canvas.drawPath(
          Path()
            ..moveTo(0, 0)
            ..lineTo(0, size.height)
            ..lineTo(size.width, size.height)
            ..lineTo(size.width, 0)
            ..addRRect(RRect.fromLTRBXY(left!, top!, right!, bottom!, round, round))
            ..close(),
          Paint()
            ..color = maskColor
            ..style = PaintingStyle.fill);

      canvas.drawPath(
          Path()
            ..addRRect(RRect.fromLTRBXY(left!, top!, right!, bottom!, round, round))
            ..close(),
          Paint()
            ..color = lineColor
            ..strokeWidth = lineWidth
            ..style = PaintingStyle.stroke);
    }
  }

  _onPadding(Size size) {
    var fw = size.width / outWidth;
    var fh = size.height / outHeight;
    if (fw > fh) {
      fw = fh;
    }
    var width = outWidth * fw / 2 - maskPadding;
    var height = outHeight * fw / 2 - maskPadding;
    centerX = size.width / 2;
    centerY = size.height / 2;
    left = centerX! - width;
    right = centerX! + width;
    top = centerY! - height;
    bottom = centerY! + height;
  }

  _onPosition() {
    if (limitations) {
      if (5 < scale) {
        scale = 5;
      }

      var fw = (right! - left!) / _image!.width;
      var fh = (bottom! - top!) / _image!.height;
      if (fw < fh) {
        fw = fh;
      }
      if (scale < fw) {
        scale = fw;
      }

      var width = _image!.width * scale / 2;
      if (left! < centerX! + drawX - width) {
        drawX = left! - centerX! + width;
      }
      if (right! > centerX! + drawX + width) {
        drawX = right! - centerX! - width;
      }

      var height = _image!.height * scale / 2;
      if (top! < centerY! + drawY - height) {
        drawY = top! - centerY! + height;
      }
      if (bottom! > centerY! + drawY + height) {
        drawY = bottom! - centerY! - height;
      }
    }
  }

  @override
  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(DiagnosticsProperty<bool>('limitations', limitations));
    properties.add(DiagnosticsProperty<bool>('isArc', isArc));
    properties.add(DoubleProperty('backBoxSize', backBoxSize));
    properties.add(ColorProperty('backBoxColor0', backBoxColor0));
    properties.add(ColorProperty('backBoxColor1', backBoxColor1));
    properties.add(ColorProperty('maskColor', maskColor));
    properties.add(ColorProperty('lineColor', lineColor));
    properties.add(DoubleProperty('lineWidth', lineWidth));
    properties.add(DoubleProperty('outWidth', outWidth));
    properties.add(DoubleProperty('outHeight', outHeight));
    properties.add(DoubleProperty('maskPadding', maskPadding));
    properties.add(DoubleProperty('round', round));
  }
}
